lib/linked_list.rb

```diff
@@ -14,136 +14,284 @@ def initialize(value, next_node = nil)
     # returns true if found, false otherwise
-    # Time Complexity:  
-    # Space Complexity
+    # Time Complexity:  O(n) where n is the number of nodes
+    # Space Complexity: O(1) 
     def search(value)
-      raise NotImplementedError
+      curr = @head
+      while curr
+        return true if curr.data == value
+        curr = curr.next
+      end
+      return false
     end
 
     # method to return the max value in the linked list
     # returns the data value and not the node
-    # Time Complexity:  
-    # Space Complexity
+    # Time Complexity:  O(n) where n is the number of nodes
+    # Space Complexity: O(1) 
     def find_max
-      raise NotImplementedError
+      return if !@head
+      curr = @head
+      max = curr.data
+      while curr
+        if max < curr.data 
+          max = curr.data
+        end
+        curr = curr.next 
+      end
+      return max
     end
 
     # method to return the min value in the linked list
     # returns the data value and not the node
-    # Time Complexity:  
-    # Space Complexity
+    # Time Complexity:  O(n) where n is the number of nodes
+    # Space Complexity: O(1) 
     def find_min
-      raise NotImplementedError
+      return if !@head
+      curr = @head
+      min = curr.data
+      while curr
+        if min > curr.data 
+          min = curr.data
+        end
+        curr = curr.next 
+      end
+      return min
     end
 
 
     # method that returns the length of the singly linked list
-    # Time Complexity:  
-    # Space Complexity
+    # Time Complexity:  O(1) (was O(n) updated class to keep track on length)
+    # Space Complexity: O(1) 
     def length
-      raise NotImplementedError
+      # count = 0
+      # curr = @head
+      # while curr && @head
+      #   curr = curr.next
+      #   count += 1
+      # end
+      # return count
+      return @length
     end
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns nil if there are fewer nodes in the linked list than the index value
-    # Time Complexity:  
-    # Space Complexity
+    # Time Complexity:  O(n) where n is the number of nodes
+    # Space Complexity: O(1) 
     def get_at_index(index)
-      raise NotImplementedError
+      target_node = node_at_index(index)
+      return if !target_node
+      return target_node.data 
     end
 
-    # method to print all the values in the linked list
-    # Time Complexity:  
-    # Space Complexity
+    def node_at_index(index)
+      length = length()
+      return if length - 1 < index ||  index < 0
+      curr = @head
+      count = 0
+      while index != count
+        curr = curr.next
+        count += 1
+      end
+      return curr
+    end
+
+    # method to return a string of all the values in the linked list
+    # Time Complexity:  O(n) where n is the number of nodes
+    # Space Complexity: O(1) 

```

Comment:
Since you're building a string of equal size to the list, you have a space complexity of O(n)